File: Sort/reorganizeString.cpp
#include <stdc++.h>

using namespace std;

class Solution {
public:
    string reorganizeString(string S) {
        int length = S.size();
        vector<int> bucket(26, 0);
        string ans = "";
        int maxVal = 0;
        int idx = 0;
        for (char word : S){
            bucket[word - 'a']++;
            if (bucket[word - 'a'] > maxVal){
                maxVal = bucket[word - 'a'];
                idx = word - 'a';
            }
        }

        if (length % 2 == 0 && maxVal > length / 2)
            return "";
        if (length % 2 == 1 && maxVal > length / 2 + 1)
            return "";

        while (bucket[idx] > 0){
            ans += (char)(idx + 'a');
        }  

        int pos = 1;
        for (int i = 0; i < 26; i++){
            while (bucket[i] > 0){
                ans.insert(pos, string(1, (char)(i + 'a')));
                pos += 2;
                bucket[i]--;
                if (pos > ans.size()){
                    pos = 1;
                }
            }
        }

        return ans;
    }
};

File: Sort/ReorganizeString.cs
using System.Text;
// 插空 + 桶排序
public class Solution {
    public string ReorganizeString(string S) {
        int length = S.Length;
        StringBuilder ans = new StringBuilder();
        int[] strs = new int[26];
        int max = 0;
        int idx = 0;
        for (int i = 0; i < length; i++){
            strs[S[i] - 'a']++;
            if (max < strs[S[i] - 'a']){
                max = strs[S[i] - 'a'];
                idx = S[i] - 'a';
            }
        }

        if (length % 2 == 0 && max > length / 2)
            return "";
        if (length % 2 == 1 && max > length / 2 + 1)
            return "";

        while (strs[idx] > 0){
            ans.Append((char)(idx + 'a'));
            strs[idx]--;
        }
        int pos = 1;
        for (int i = 0; i < 26; i++){
            while (strs[i] > 0){
                ans.Insert(pos, (char)(i + 'a'));
                pos += 2;
                strs[i]--;
                if (pos > ans.Length)
                    pos = 1;
            }
        }

        return ans.ToString();
    }
}